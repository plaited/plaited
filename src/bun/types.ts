import type {
  BPEvent,
  Actions,
  Trigger,
  BThreads,
  BSync,
  BThread,
  UseSnapshot,
  UseFeedback,
} from '../behavioral.js'
import { Attrs, TemplateObject } from '../jsx/jsx.types.js';

export type ServerWebSocket<T = unknown> = Parameters<
  Extract<Parameters<typeof Bun.serve<T>>[0], { websocket: unknown; unix?: never }>['websocket']['message']
>['0']

export type DefineSocketHandler<T> = (args: {
  send: (address: string) => (event: BPEvent<string>) => ReturnType<ServerWebSocket['send']>
  trigger: Trigger
  useSnapShot: UseSnapshot
  useFeedback: UseFeedback
  bThreads: BThreads
  bSync: BSync
  bThread: BThread
  getContext: () => T
}) => Actions

export type DefineSocketConfig = { publicEvents: string[]; address: string }

export type DefineSocketReturn<T> = {
  address: string
  open(ws: ServerWebSocket<T>): void
  message: (ws: ServerWebSocket<T>, event: BPEvent) => void
}

export type MiddleWareHandler = <T extends Record<string, unknown> = Record<string, unknown>>(
  req: Request,
  ctx?: T & { next: Handler },
) => Promise<Response>

export type Handler = <T extends Record<string, unknown> = Record<string, unknown>>(
  req: Request,
  ctx?: T,
) => Promise<Response>

export type AjaxPageData = [Document, Record<string, Element | string>]


export type Story<T = Attrs> = {
  render: (attrs: T) => TemplateObject
  play: (attrs: T) => void
  description?: string
  a11y: []
  attrs: T
}