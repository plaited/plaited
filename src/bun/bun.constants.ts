export const IGNORED_DIR_PREFIX = '[!_]'
export const GLOB_PATTERN_WORKERS = `**/${IGNORED_DIR_PREFIX}*/worker.@(tsx|ts)`
export const GLOB_PATTERN_MODULES = `**/${IGNORED_DIR_PREFIX}*/module.@(tsx|ts)`
