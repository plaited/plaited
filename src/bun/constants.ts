export const IGNORED_DIR_PREFIX = '[!_]'
export const GLOB_PATTERN_WORKER = `**/${IGNORED_DIR_PREFIX}*/worker.@(tsx|ts)`
export const GLOB_PATTERN_ROUTE = `**/${IGNORED_DIR_PREFIX}*/route.@(tsx|ts)`
export const GLOB_PATTERN_STORIES = `**/*.stories.@(tsx|ts)`