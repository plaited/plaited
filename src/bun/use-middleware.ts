import type { Handler, MiddleWareHandler } from './bun.types.js'

export const useMiddleware = (middleware?: MiddleWareHandler) => {
  return async (handler: Handler) => {
    if (!middleware) return (req: Request, ctx: Record<string, string>) => handler(req, ctx)
    return async (req: Request, ctx: Record<string, string>) => {
      const resp = await middleware(req, {
        ...ctx,
        next: () => handler(req, ctx),
      })
      return resp
    }
  }
}
