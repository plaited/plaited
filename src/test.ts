export * from './test/assert.types.js'
export * from './test/assert.constants.js'
export { assert } from './test/assert.js'
export { findByAttribute } from './test/find-by-attribute.js'
export { findByText } from './test/find-by-text.js'
export { fireEvent } from './test/fire-event.js'
export { match } from './test/match.js'
export { throws } from './test/throws.js'
export { wait } from './utils/wait.js'
export { AssertionError, MissingTestParamsError, TimeoutError } from './test/errors.js'
export { usePlay } from './test/use-play.js'
