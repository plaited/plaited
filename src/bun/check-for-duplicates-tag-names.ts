import type { PlaitedTemplate } from '../client/types.js'
import { isTypeOf } from '../utils/true-type-of.js'
import { PLAITED_TEMPLATE_IDENTIFIER } from '../shared/constants.js'
import { GLOB_PATTERN_DUPLICATE_TAGS } from './bun.constants.js'

const isPlaitedComponent = (mod: unknown): mod is PlaitedTemplate =>
  isTypeOf<PlaitedTemplate>(mod, 'function') && mod?.$ === PLAITED_TEMPLATE_IDENTIFIER

export const checkForDuplicateTagNames = async (dir: string, pattern = GLOB_PATTERN_DUPLICATE_TAGS) => {
  const glob = new Bun.Glob(pattern)
  const components = new Map<string, string[]>()
  const componentPaths = await Array.fromAsync(glob.scan({ cwd: dir }))
  await Promise.all(
    componentPaths.map(async (path) => {
      try {
        const modules = await import(path)
        for (const name of modules) {
          const mod = modules[name]
          if (!mod || !isPlaitedComponent(mod)) continue
          if (components.has(mod.tag)) {
            components.get(mod.tag)?.push(path)
          } else {
            components.set(mod.tag, [path])
          }
        }
        return path
      } catch (err) {
        console.error(err)
      }
    }),
  )
  return [...components].filter(([_, paths]) => paths.length > 1)
}
