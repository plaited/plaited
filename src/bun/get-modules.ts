import { GLOB_PATTERN_MODULES } from './bun.constants.js'
import { isPlaitedTemplate } from './is-plaited-template.js'

const transpiler = new Bun.Transpiler({
  loader: 'tsx',
})

const registry = new Map<string, { filePath: string; exportName: string }>()

const getImports = async (templatePaths: { path: string; kind: string }[]) => {
  const imports = new Set<string>()
  await Promise.all(
    templatePaths.map(async ({ path: filePath }) => {
      try {
        const modules = await import(filePath)
        for (const exportName of modules) {
          const mod = modules[exportName]
          if (!mod || !isPlaitedTemplate(mod)) continue
          if (registry.has(mod.tag)) {
            console.error(
              `Error ${mod.tag} is already registered ${registry.get(mod.tag)?.exportName}@${registry.get(mod.tag)}\n Change tag name of ${exportName}@${filePath}`,
            )
            return
          } else {
            registry.set(mod.tag, { filePath, exportName })
            imports.add(filePath)
          }
        }
      } catch (err) {
        console.error(err)
      }
    }),
  )
  return imports
}

export const getModules = async (dir: string) => {
  const glob = new Bun.Glob(GLOB_PATTERN_MODULES)
  const storyPaths = await Array.fromAsync(glob.scan({ cwd: dir }))
  registry.clear()
  const entries = await Promise.all(
    storyPaths.map(async (filePath) => {
      try {
        const file = Bun.file(filePath)
        const code = await file.text()
        const result = transpiler.scanImports(code)
        const imports = await getImports(result)
        let fileImports = ''
        for (const entry of imports) {
          fileImports += `import '${entry}'\n`
        }
        const regex = /\.tsx?$/
        return { [filePath.replace(regex, '.plaited')]: fileImports }
      } catch (err) {
        console.error(err)
      }
    }),
  )
  return entries
}
