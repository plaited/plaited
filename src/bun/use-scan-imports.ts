import { dirname, sep, resolve } from 'node:path'
import { TEMPLATES_DIR } from './bun.constants.js'

export const useScanImports = (dir: string = TEMPLATES_DIR) => {
  const transpiler = new Bun.Transpiler({
    loader: 'tsx',
  })
  return (filePath: string) => {
    const importPaths = transpiler
      .scanImports(filePath)
      .flatMap(({ path }) => (path.split(sep).includes(dir) ? [resolve(dirname(filePath), path)] : []))
    return importPaths
  }
}
