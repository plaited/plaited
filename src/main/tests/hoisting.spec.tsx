import { test, expect } from 'bun:test'
import { type TemplateObject, defineElement, css } from 'plaited'
import beautify from 'beautify'

const render = (tpl: TemplateObject) => beautify(tpl.html.join(''), { format: 'html' })

const styles = css.create({
  nestedLabel: {
    fontWeight: 'bold',
  },
  nestedComponent: {
    display: 'flex',
    flexDirection: 'column',
  },
  slottedParagraph: {
    color: 'rebeccapurple',
  },
  topComponent: {
    display: 'block',
  },
  image: {
    width: '100%',
    aspectRatio: '16 / 9',
  },
})

const NestedCustomElement = defineElement({
  tag: 'nested-component',
  shadowDom: (
    <>
      <span
        {...styles.nestedLabel}
        part='span'
      >
        inside nested template
      </span>
      <slot name='nested' />
    </>
  ),
})

test('createTemplate: CustomElement hoisting its styles and parts', () => {
  const el = <NestedCustomElement />
  expect({
    content: render(el),
    stylesheets: el.stylesheets,
    parts: el.parts,
  }).toMatchSnapshot()
})

test('createTemplate: CustomElement with declarative shadow dom & hoist styles and part', () => {
  const el = (
    <NestedCustomElement
      {...styles.nestedComponent}
      part='nested-component S1'
    />
  )
  expect({
    content: render(el),
    stylesheets: el.stylesheets,
    parts: el.parts,
  }).toMatchSnapshot()
})

test('createTemplate: CustomElement with styled slotted component', () => {
  const el = (
    <NestedCustomElement {...styles.slottedParagraph}>
      <p
        slot='nested'
        part='S2 paragraph'
      >
        slotted paragraph
      </p>
    </NestedCustomElement>
  )
  expect({ content: render(el), stylesheets: el.stylesheets, parts: el.parts }).toMatchSnapshot()
})

const TopCustomElement = defineElement({
  tag: 'top-component',
  shadowDom: (
    <NestedCustomElement {...styles.slottedParagraph}>
      <p
        slot='nested'
        part='paragraph S2'
      >
        slotted paragraph
      </p>
    </NestedCustomElement>
  ),
})

test('createTemplate: CustomElement with declarative shadow dom and nested declarative shadow dom', () => {
  const el = <TopCustomElement part='S3' />
  expect({ content: render(el), stylesheets: el.stylesheets, parts: el.parts }).toMatchSnapshot()
})

test('createTemplate: CustomElement with declarative shadow dom and nested declarative shadow dom plus host styles', () => {
  const el = <TopCustomElement {...styles.topComponent} />
  expect({
    content: render(el),
    stylesheets: el.stylesheets,
    parts: el.parts,
  }).toMatchSnapshot()
})

test('createTemplate: CustomElement with declarative shadow dom and nested declarative shadow dom plus host styles and child', () => {
  const el = (
    <TopCustomElement {...styles.topComponent}>
      <img {...styles.image} />
    </TopCustomElement>
  )
  expect({ content: render(el), stylesheets: el.stylesheets, parts: el.parts }).toMatchSnapshot()
})

const hoistStyles = css.create({
  var1: {
    width: '100%',
  },
  var2: {
    width: '100%',
  },
  var3: {
    color: 'blue',
  },
})

test('ssr: Properly hoist and deduplicates multiple stylesheets on a single node', () => {
  expect((<div {...css.assign(hoistStyles.var1, hoistStyles.var2, hoistStyles.var3)} />).stylesheets.length).toBe(2)
})

test('ssr: filters out falsey style object', () => {
  expect((<div {...css.assign(hoistStyles.var1, hoistStyles.var2, false, undefined, null)} />).stylesheets.length).toBe(
    1,
  )
})

test('ssr: Properly deduplicates duplicate part identifiers', () => {
  expect((<div part='S1 div S1' />).parts.length).toBe(2)
})
