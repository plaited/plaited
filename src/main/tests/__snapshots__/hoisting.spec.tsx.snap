// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`createTemplate: CustomElement hoisting its styles and parts 1`] = `
{
  "content": 
"<nested-component exportparts="span:nested-component--span,">
    <template shadowrootmode="open" shadowrootdelegatesfocus>
        <style>
            .p_ju9u1n{font-weight:bold;}
        </style><span class="nestedLabel__15w1ydc p_ju9u1n" part="span">inside nested template</span>
        <slot
        name="nested"></slot>
    </template>
</nested-component>"
,
  "parts": [],
  "stylesheets": [],
}
`;

exports[`createTemplate: CustomElement with declarative shadow dom & hoist styles and part 1`] = `
{
  "content": 
"<nested-component class="nestedComponent_ey8y76 p1rlu8mi p_8mez4t" exportparts="span:nested-component--span,"
part="nested-component S1">
    <template shadowrootmode="open" shadowrootdelegatesfocus>
        <style>
            .p_ju9u1n{font-weight:bold;}
        </style><span class="nestedLabel__15w1ydc p_ju9u1n" part="span">inside nested template</span>
        <slot
        name="nested"></slot>
    </template>
</nested-component>"
,
  "parts": [
    "nested-component",
    "S1",
  ],
  "stylesheets": [
    ".p1rlu8mi{display:flex;}",
    ".p_8mez4t{flex-direction:column;}",
  ],
}
`;

exports[`createTemplate: CustomElement with styled slotted component 1`] = `
{
  "content": 
"<nested-component class="slottedParagraph_g28n4j p91m869" exportparts="span:nested-component--span,">
    <template shadowrootmode="open" shadowrootdelegatesfocus>
        <style>
            .p_ju9u1n{font-weight:bold;}
        </style><span class="nestedLabel__15w1ydc p_ju9u1n" part="span">inside nested template</span>
        <slot
        name="nested"></slot>
    </template>
    <p slot="nested" part="S2 paragraph">slotted paragraph</p>
</nested-component>"
,
  "parts": [
    "S2",
    "paragraph",
  ],
  "stylesheets": [
    ".p91m869{color:rebeccapurple;}",
  ],
}
`;

exports[`createTemplate: CustomElement with declarative shadow dom and nested declarative shadow dom 1`] = `
{
  "content": 
"<top-component exportparts="paragraph:top-component--paragraph, S2:top-component--S2,"
part="S3">
    <template shadowrootmode="open" shadowrootdelegatesfocus>
        <style>
            .p91m869{color:rebeccapurple;}
        </style>
        <nested-component class="slottedParagraph_g28n4j p91m869" exportparts="span:nested-component--span,">
            <template shadowrootmode="open" shadowrootdelegatesfocus>
                <style>
                    .p_ju9u1n{font-weight:bold;}
                </style><span class="nestedLabel__15w1ydc p_ju9u1n" part="span">inside nested template</span>
                <slot
                name="nested"></slot>
            </template>
            <p slot="nested" part="paragraph S2">slotted paragraph</p>
        </nested-component>
    </template>
</top-component>"
,
  "parts": [
    "S3",
  ],
  "stylesheets": [],
}
`;

exports[`createTemplate: CustomElement with declarative shadow dom and nested declarative shadow dom plus host styles 1`] = `
{
  "content": 
"<top-component class="topComponent_46img7 p131mv1i" exportparts="paragraph:top-component--paragraph, S2:top-component--S2,">
    <template shadowrootmode="open" shadowrootdelegatesfocus>
        <style>
            .p91m869{color:rebeccapurple;}
        </style>
        <nested-component class="slottedParagraph_g28n4j p91m869" exportparts="span:nested-component--span,">
            <template shadowrootmode="open" shadowrootdelegatesfocus>
                <style>
                    .p_ju9u1n{font-weight:bold;}
                </style><span class="nestedLabel__15w1ydc p_ju9u1n" part="span">inside nested template</span>
                <slot
                name="nested"></slot>
            </template>
            <p slot="nested" part="paragraph S2">slotted paragraph</p>
        </nested-component>
    </template>
</top-component>"
,
  "parts": [],
  "stylesheets": [
    ".p131mv1i{display:block;}",
  ],
}
`;

exports[`createTemplate: CustomElement with declarative shadow dom and nested declarative shadow dom plus host styles and child 1`] = `
{
  "content": 
"<top-component class="topComponent_46img7 p131mv1i" exportparts="paragraph:top-component--paragraph, S2:top-component--S2,">
    <template shadowrootmode="open" shadowrootdelegatesfocus>
        <style>
            .p91m869{color:rebeccapurple;}
        </style>
        <nested-component class="slottedParagraph_g28n4j p91m869" exportparts="span:nested-component--span,">
            <template shadowrootmode="open" shadowrootdelegatesfocus>
                <style>
                    .p_ju9u1n{font-weight:bold;}
                </style><span class="nestedLabel__15w1ydc p_ju9u1n" part="span">inside nested template</span>
                <slot
                name="nested"></slot>
            </template>
            <p slot="nested" part="paragraph S2">slotted paragraph</p>
        </nested-component>
    </template>
    <img class="image__c0ru3k p_mygp1 p_x8x9vx" />
</top-component>"
,
  "parts": [],
  "stylesheets": [
    ".p131mv1i{display:block;}",
    ".p_mygp1{width:100%;}",
    ".p_x8x9vx{aspect-ratio:16 / 9;}",
  ],
}
`;
